### PR TITLE
[ipgen,flash_ctrl] Create ipgen files

### PR DESCRIPTION
**Pull request by matutem**
Create the flash_ctrl generated files using ipgen, but keep using the legacy flow for now. This is so the full transition is broken into two stages for ease of reviewing. A subsequent change will complete the transition to ipgen and remove the legacy files.

Just like previous IP transitions to ipgen, most of legacy files are simply copied to hw/ip_templates/flash_ctrl, but the following changes are noteworthy:

- The third commit moves .tpl files to the locations needed for ipgen, and wipes out from ip_templates the files that are generated so the tree will no longer have these useless and confusing files.
- The fourth commit makes additional files into templates since their generated text contains the name of top.
- The fifth commit is needed since ipgen uses plain dictionaries, whereas the old flow uses some class objects.
- The sixth commit fixes core files that build templated files to use instance_vlnv and declare a virtual target, and changes other core files to depend on their virtual name instead.
- The seventh commit makes azure avoid building under the ip_templates tree since build should only happen in the target generated trees.

### CODE REVIEW

File: hw/ip_templates/flash_ctrl/rtl/flash_ctrl.sv [hunk 1]
```diff
@@ -0,0 +1,1498 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Flash Controller Module
```
Comment (msfschaffner) [inline hw/ip_templates/flash_ctrl/rtl/flash_ctrl.sv:5]: Is this file still needed here or should it be removed?

**Review by msfschaffner — COMMENTED** [review]
Thanks for doing this @matutem. Do I understand correctly that the build still uses the old files at the moment, and you are basically going to change the fusesoc dependency to the new ipgen'd files in a follow up PR?

Does it make sense to diff the RTL files (and potentially DV files) to make sure they remained identical during this operation?

**Comment by matutem** [comment]
@msf, below are the diffs of the tpl files. Notice flash_ctrl_pkg.sv has an edit for the sake of simplicity: we use already declared variables instead of recreating them from constants.

There are many required differences in the doc files, since they contain many paths that need to change.

 diff hw/top_earlgrey/ip/flash_ctrl/rtl/autogen/flash_ctrl_pkg.sv hw/top_earlgrey/ip_autogen/flash_ctrl/rtl/flash_ctrl_pkg.sv 
5,11c5
< // ------------------- W A R N I N G: A U T O - G E N E R A T E D   C O D E !! -------------------//
< // PLEASE DO NOT HAND-EDIT THIS FILE. IT HAS BEEN AUTO-GENERATED WITH THE FOLLOWING COMMAND:
< // Copyright lowRISC contributors.
< // Licensed under the Apache License, Version 2.0, see LICENSE for details.
< // SPDX-License-Identifier: Apache-2.0
< //
< // Flash Controller module.
---
> // Flash Controller package.
31,33c25,27
<     10,
<     1,
<     2
---
>     flash_ctrl_reg_pkg::NumInfos0,
>     flash_ctrl_reg_pkg::NumInfos1,
>     flash_ctrl_reg_pkg::NumInfos2
35,39c29
<   parameter int InfosPerBank    = max_info_pages('{
<     10,
<     1,
<     2
<   });
---
>   parameter int InfosPerBank    = max_info_pages(InfoTypeSize);
78,82d67
<   //parameter logic [PageW-1:0] InfoPartitionEndAddr [InfoTypes] = '{
<   //  9,
<   //  0,
<   //  1
<   //};
$ diff hw/top_earlgrey/ip/flash_ctrl/rtl/autogen/flash_ctrl.sv hw/top_earlgrey/ip_autogen/flash_ctrl/rtl/flash_ctrl.sv 
5,10d4
< // ------------------- W A R N I N G: A U T O - G E N E R A T E D   C O D E !! -------------------//
< // PLEASE DO NOT HAND-EDIT THIS FILE. IT HAS BEEN AUTO-GENERATED WITH THE FOLLOWING COMMAND:
< // Copyright lowRISC contributors.
< // Licensed under the Apache License, Version 2.0, see LICENSE for details.
< // SPDX-License-Identifier: Apache-2.0
< //
$ diff hw/top_earlgrey/ip/flash_ctrl/data/autogen/flash_ctrl.hjson hw/top_earlgrey/ip_autogen/flash_ctrl/data/flash_ctrl.hjson 
4,9d3
< //
< // ------------------- W A R N I N G: A U T O - G E N E R A T E D   C O D E !! -------------------//
< // PLEASE DO NOT HAND-EDIT THIS FILE. IT HAS BEEN AUTO-GENERATED WITH THE FOLLOWING COMMAND:
< // Copyright lowRISC contributors.
< // Licensed under the Apache License, Version 2.0, see LICENSE for details.
< // SPDX-License-Identifier: Apache-2.0
$ diff hw/top_earlgrey/ip/flash_ctrl/rtl/autogen/flash_ctrl_region_cfg.sv hw/top_earlgrey/ip_autogen/flash_ctrl/rtl/flash_ctrl_region_cfg.sv 
5,10d4
< // ------------------- W A R N I N G: A U T O - G E N E R A T E D   C O D E !! -------------------//
< // PLEASE DO NOT HAND-EDIT THIS FILE. IT HAS BEEN AUTO-GENERATED WITH THE FOLLOWING COMMAND:
< // Copyright lowRISC contributors.
< // Licensed under the Apache License, Version 2.0, see LICENSE for details.
< // SPDX-License-Identifier: Apache-2.0
< //

File: hw/ip_templates/flash_ctrl/rtl/flash_ctrl.sv (hunk 1 above)
Comment (matutem) [reply hw/ip_templates/flash_ctrl/rtl/flash_ctrl.sv]: Good catch, done.

**Comment by msfschaffner** [comment]
Thanks for the diff - this all looks reasonable, thanks!

**Review by msfschaffner — APPROVED** [review]
